executor/cook/executor.py

```diff
@@ -0,0 +1,364 @@
+        :return: The id of the task.
+    """
+    return task['task_id']['value']
+
+
+def create_status(task_id, task_state):
+    """
+        :param task_id: The task id.
+        :param task_state: The state of the task to report.
+        :return: a status dictionary that can be sent to the driver.
+    """
+    return {'task_id': {'value': task_id},
+            'state': task_state,
+            'timestamp': time.time()}
+
+
+def update_status(driver, task, task_state):
+    """
+        :param driver: The driver to send the status update to.
+        :param task: The task whose status update to send.
+        :param task_state: The state of the task which will be sent to the driver.
+        :return: Nothing.
+    """
+    task_id = get_task_id(task)
+    logging.debug('Updating task {} state to {}'.format(task_id, task_state))
+    status = create_status(task_id, task_state)
+    driver.sendStatusUpdate(status)
+
+
+def send_message(driver, message, max_message_length):
+    """
+        :param driver: The driver to send the message to.
+        :param message: The raw message to send.
+        :param max_message_length: The allowed max message length after encoding
+        :return: whether the message was successfully sent
+    """
+    logging.debug('Sending framework message {}'.format(message))
+    encoded_message = encode_data(str(message).encode('utf8'))
+    if len(encoded_message) < max_message_length:
+        driver.sendFrameworkMessage(encoded_message)
+        return True
+    else:
+        log_message_template = 'Unable to send message {} as it exceeds allowed max length of {}'
+        logging.info(log_message_template.format(message, max_message_length))
+        return False
+
+
+def launch_task(driver, task, stdout_file, stderr_file):
+    """
+        :param driver: The mesos driver to use.
+        :param task: The task to execute.
+        :param stdout_file: The file to use to redirect stdout.
+        :param stderr_file: The file to use to redirect stderr.
+        :return: The tuple containing the process launched, stdout file, and stderr file.
+    """
+    try:
+        update_status(driver, task, 'TASK_RUNNING')
+
+        command = decode_data(task['data']).decode('utf8')
+        logging.info('Command: {}'.format(command))
+        stdout = open(stdout_file, 'a+')
+        stderr = open(stderr_file, 'a+')
+        process = subprocess.Popen(command, shell=True, stdout=stdout, stderr=stderr)
+
+        return process, stdout, stderr
+    except Exception:
+        logging.exception('Error in launch_task')
+        update_status(driver, task, 'TASK_FAILED')
+
+
+def is_running(process):
+    """
+        :param process: The process to query
+        :return: whether the process is still running.
+    """
+    return process.poll() is None
+
+
+def has_crashed(process):
+    """
+        :param process:
+        :return:
+    """
+    return not is_running(process) and process.poll() is not 0
```

Comment:
`is_running` is based upon the result of `process.poll()`.  Should we not call `process.poll()` directly inside `has_crashed` to avoid calling it swiace?